services/web/server/src/simcore_service_webserver/meta_handlers.py

```diff
@@ -0,0 +1,277 @@
+
+IterationTuple = Tuple[ProjectID, CommitID]
+
+
+async def list_project_iterations(
+    vc_repo: VersionControlForMetaModeling,
+    project_uuid: ProjectID,
+    commit_id: CommitID,
+    offset: int = 0,
+    limit: Optional[int] = None,
+) -> Tuple[List[IterationTuple], int]:
+
+    assert offset >= 0  # nosec
+
+    repo_id = await vc_repo.get_repo_id(project_uuid)
+    assert repo_id is not None
+
+    total_number_of_iterations = 0  # count number of iterations
+
+    # Search all subsequent commits (i.e. children) and retrieve their tags
+    # Select range on those tagged as iterations and returned their assigned workcopy id
+    # Can also check all associated project uuids and see if they exists
+    # FIXME: do all these operations in database
+    tags_per_child: List[List[TagProxy]] = await vc_repo.get_children_tags(
+        repo_id, commit_id
+    )
+
+    iterations: List[Tuple[ProjectID, CommitID]] = []
+    for n, tags in enumerate(tags_per_child):
+
+        # FIXME: the db query did not guarantee
+        # not sorted
+        # not limited
+        # not certain if tags we need
+        try:
+            iteration: Optional[ProjectIteration] = None
+            workcopy_id: Optional[ProjectID] = None
+
+            for tag in tags:
+                if pim := ProjectIteration.from_tag_name(
+                    tag.name, return_none_if_fails=True
+                ):
+                    if iteration:
+                        raise ValueError(
+                            f"This entry has more than one iteration {tag=}"
+                        )
+                    iteration = pim
+                elif pid := parse_workcopy_project_tag_name(tag.name):
+                    if workcopy_id:
+                        raise ValueError(
+                            f"This entry has more than one workcopy  {tag=}"
+                        )
+                    workcopy_id = pid
+                else:
+                    log.debug("Got %s for children of %s", f"{tag=}", f"{commit_id=}")
+
+            if not workcopy_id:
+                raise ValueError("No working copy found")
+            if not iteration:
+                raise ValueError("No iteration tag found")
+
+            iterations.append((workcopy_id, iteration.iter_index))
+
+        except ValueError as err:
+            log.debug("Skipping child %s: %s", n, err)
+
+    total_number_of_iterations = len(iterations)
+
+    # sort and select. If requested interval is outside of range, it returns empty
+    iterations.sort(key=lambda tup: tup[1])
+
+    if limit is None:
+        return iterations[offset:], total_number_of_iterations
+
+    return iterations[offset : (offset + limit)], total_number_of_iterations
+
+
+async def create_or_get_project_iterations(
+    vc_repo: VersionControlForMetaModeling,
+    project_uuid: ProjectID,
+    commit_id: CommitID,
+) -> List[IterationTuple]:
+
+    raise NotImplementedError()
```

Comment:
is this a forgotten one? or will it be implemented? maybe a little comment here would help if that is the case

Reply:
Added link to TODO list in #2735